Doc/library/functions.rst

```diff
@@ -496,44 +496,38 @@ are always available.  They are listed here in alphabetical order.
 
 .. function:: eval(expression[, globals[, locals]])
 
-   The arguments are a string and optional globals and locals.  If provided,
-   *globals* must be a dictionary.  If provided, *locals* can be any mapping
-   object.
-
-   The *expression* argument is parsed and evaluated as a Python expression
-   (technically speaking, a condition list) using the *globals* and *locals*
-   dictionaries as global and local namespace.  If the *globals* dictionary is
-   present and does not contain a value for the key ``__builtins__``, a
-   reference to the dictionary of the built-in module :mod:`builtins` is
-   inserted under that key before *expression* is parsed.  That way you can
-   control what builtins are available to the executed code by inserting your
-   own ``__builtins__`` dictionary into *globals* before passing it to
-   :func:`eval`.  If the *locals* dictionary is omitted it defaults to the
-   *globals* dictionary.  If both dictionaries are omitted, the expression is
-   executed with the *globals* and *locals* in the environment where
-   :func:`eval` is called.  Note, *eval()* does not have access to the
-   :term:`nested scopes <nested scope>` (non-locals) in the enclosing
-   environment.
-
-   The return value is the result of
-   the evaluated expression. Syntax errors are reported as exceptions.  Example:
+   This function supports the dynamic evaluation of Python expression.  The
+   first argument can be a string or a code object.  The optional arguments
+   specify the globals and locals respectively.  If provided, *globals* must be
+   a dictionary.  If provided, *locals* can be any mapping object.
```

Comment:
What?? In the source for `eval/exec` in Python/bltinmodule.c, the locals are checked by `PyMapping_Check()` while globals checked by `PyDict_Check()`. I'm not sure why the exception was about "'int' object is not callable" -- it seems the name `a` did get resolved in the locals through the `Counter` object, but somehow there was an action to the effect of `1()`. This might be a bug! Right now I don't have enough spoons to track it down, but I should be verifying it and maybe submitting a bug report soon.

Reply:
This works as intended:
```python
>>> ct = Counter("abcd")
>>> eval("a", {}, ct)
1
```
But not this:
```python
>>> eval("print(a)", {}, ct)
```
Expected return value of `eval` is `None` (return value of `print` function), but the cryptic exception is raised.

Reply:
Also, this:
```python
>>> eval("print(1)", {}, ct)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
TypeError: 'int' object is not callable
```

Reply:
I see! The locals has the highest priority in name resolution. The `Counter` instance (`ct` in the above example) behaves as if any key is in it (with initial value 0). When the machinery underlying `eval` does its work, it first looks up the name `print` in locals. Unfortunately it gets what it wishes for: the count for the key `"print"` in the counter `ct`, which is the `int` instance `0`. Then `0` gets called as the function.  The evaluation boils down to `0(1)` which raises `TypeError`.

Reply:
Awesome! Thank you for investigating this! I learnt a lot :).

Reply:
Thanks, me too. This is the weirdest thing I've seen in Python in a while.